doc/arch/adr-005-api-client-app-auth.md

```diff
@@ -0,0 +1,108 @@
+
+We're creating a [proper, publicly documented API](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/geoprocessing_api/views.py) around our analyze and 
+RWD endpoints: the [Geoprocessing API.](https://github.com/WikiWatershed/model-my-watershed/blob/develop/doc/arch/adr-004-geoprocessing-api.md) We want to be able to keep track of 
+the API's users so that we can prevent them from sending off too many
+requests, revoke the access of problematic users, and some day add a 
+paid tier of use. To accomplish this we've decided all API requests
+should be authenticated via token. Our client app, however, also consumes
+the API, and we still want "guest" users of the app to be able to analyze
+shapes and delineate watersheds. We need a way to identify the
+client app as a special case that the app should allow anonymous requests 
+from.
+
+Because everything we send to and from the client is exposed to the app user,
+any mechanism we give the client app to identify itself will be fairly
+easy to uncover and then fake by anyone trying to get un-checked access to the
+API. Our solution should:
+
+- Allow us to easily update the client "secret" so that whenever someone steals it, we can
+just cycle it out to a new one. (Or this could happen at some regular interval)
+- Not be overly complicated/difficult for us when we know there's no truly secure
+solution (barring doing all our rendering server-side)
+
+Some possibilities are:
+
+1. Give each environment of the client app its own API token with which it
+would make guest user requests
+   - Keeps a single system for API authentication
+   
+1. Set a special flag in the session token that the API would check for in
+unauthenticated requests
+   - Could be the CSRF token or something else. Might involve some custom
+   generation/management
+   
+1. Some combination of (1) and (2)
+   - Things could get a lot more complex for us, but not that much more complex
+   for any would-be attacker
+
+## Decision
+
+**Give the client app its own API token**
+
+Setting up the app with its own API token will allow us to both
+- cycle the key easily because Django Rest Framework (DRF) handles token
+generation/re-generation, and we already use it for "real" API users
+- keep the auth system simple: everyone uses a token for auth
+
+Starting with a client API token as a base, we can always add additional checks on
+top as enhancements down the road.
+
+##### How the client app should get the token
+
+If we hard-code the token on the frontend (or anywhere), we'll have to deploy any time we need
+to cycle the token. We should instead pass the token to the client app via `clientSettings`. 
+
+##### How the server should get the token
+
+To get the token to put in the client's settings, the server can look up the
+token via the client app's `user_id`. The `user_id` will have to remain
+secret so that there's no programmatic way get the token outside of swiping
+it off the client app.
```

Comment:
I may be missing something, but where is the token stored?

Reply:
The token is stored in the `authtoken_token` table created by [`rest_framework.authtoken`](http://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication)